sunpy/net/helioviewer.py

```diff
@@ -23,44 +23,86 @@
         return self._get_json(params)
-
-    def get_closest_image(self, date, **kwargs):
+    
+    def get_source_id(self):
+        """
+        Returns a dictionary containing the source ID of all the
+        images in the data source
+        """
+        d = dict()
+        from sunpy.net.helioviewer import HelioviewerClient
+        hv = HelioviewerClient()  # doctest: +REMOTE_DATA
+        datasources = hv.get_data_sources() # doctest: +REMOTE_DATA
+        for name, observatory in datasources.items():  # doctest: +REMOTE_DATA
+            if name == "TRACE":  # doctest: +REMOTE_DATA
+                for instrument, params in observatory.items():  # doctest: +REMOTE_DATA
+                    d[(name, instrument)] = params['sourceId']  # doctest: +REMOTE_DATA
+            else:  # doctest: +REMOTE_DATA
+                for inst, detectors in observatory.items():  # doctest: +REMOTE_DATA
+                    for wavelength, params in detectors.items():  # doctest: +REMOTE_DATA
+                        if name in ["Hinode", "STEREO_A", "STEREO_B"] or wavelength in ["C2", "C3"]:  # doctest: +REMOTE_DATA
+                            for wave, adict in params.items():  # doctest: +REMOTE_DATA
+                                d[(name, inst, wavelength, wave)] = adict['sourceId']  # doctest: +REMOTE_DATA
+                        else:  # doctest: +REMOTE_DATA
+                            d[(name, inst, wavelength)] = params['sourceId']  # doctest: +REMOTE_DATA
+
+        return(d)
+
+    def display_source_id(self):
+        """
+        Function to display the source ID of all the images
+        in the data source.
+        """ 
+        from sunpy.net.helioviewer import HelioviewerClient
+        hv = HelioviewerClient()  # doctest: +REMOTE_DATA
+        d_source_id = hv.get_source_id() # doctest: +REMOTE_DATA
+        print("{" + "\n".join("{}: {}".format(k, v) for k, v in d_source_id.items()) + "}")
+
+    def get_closest_image(self, date, sourceid):
         """Finds the closest image available for the specified source and date.
 
         For more information on what types of requests are available and the
         expected usage for the response, consult the Helioviewer API
-        documentation: http://legacy.helioviewer.org/api/docs/v1/ .
+        documentation: http://api.helioviewer.org/docs/v2/ .
 
         Parameters
         ----------
         date : `datetime.datetime`, `str`
             A string or datetime object for the desired date of the image
-        observatory : string
-            (Optional) Observatory name
-        instrument : string
-            (Optional) instrument name
-        detector : string
-            (Optional) detector name
-        measurement : string
-            (Optional) measurement name
-        sourceId : int
-            (Optional) data source id
+        sourceid : int
+            It is a unique number that can be used to refer a particular image
+            datasource in the API reqeust.
```

Comment:
We want to go back to the older API now and not use sourceid directly. So you will have to make those changes!

Reply:
But still, we need to tell them that the source ID is a necessary parameter. Though we are not asking them to input any number still it is a required parameter.

Reply:
We decided to go back to the older API and abstract out the sourceid. The user should be told about it as in the example but not use it directly.

Reply:
But if I remove it from a necessary parameter then the whole function will change

Reply:
Yeah, thats we ended up deciding upon. We go back to the older API and the users don't have to use the sourceid directly.

Reply:
Oh ok. It will take me some time. Actually my exams will be starting soon so I'll be a bit slow

Reply:
That is fine. Your exams take priority and we don't tend to code to a deadline, so PRs are allowed to take as long as they need to.

Reply:
I need some clarification. Version 2 demands the user to enter source iD if they want to download an image. We have to pass source ID in some way. So according to me, we have to keep source ID as a required parameter. Yes, we can add observatory, instrument etc as required parameters as well and then calculate the source ID using those parameters but I think we cannot remove the source ID from parameter field

Reply:
But the helioviewer API takes its values from the function. If we won't be having source ID as a parameter how will we pass source ID to the API

Reply:
The way we do it right now?

Reply:
Oh now I get it. Thanks that cleared my confusion

Reply:
One more doubt. How will we make source ID a compulsory argument? Earlier it was optional. We can use the function like before but how will me make it compulsory?

Reply:
We won't be making it a compulsory argument.

Reply:
But then if a user don't passes the source ID Helioviewer won't download the image

Reply:
The user doesn't pass in sourceid, our function will.

Reply:
But our function is not taking source ID as an argument

Reply:
Exactly. We work out the sourceid based on the user input and then pass it on.

Reply:
I am still not very confident about it but I will try

Reply:
That's fine, we can help.

Reply:
Shall we create two different functions? One that will take input from the user like observatory etc and another that will take all these values and download the image?

Reply:
I don't see the need to create two different functions.

Reply:
Oh ok

Reply:
If we use this function 
filepath = hv.download_jp2('2012/07/03 14:30:00', observatory='SDO', instrument='AIA', detector='AIA', measurement='171') 
how will we pass source ID because as soon as this function is called an image will be downloaded. I can calculate the source ID but I am not clear how will you pass it if it is not a required parameter

Reply:
You don't pass it into `download_jp2`. You use `observatory='SDO', instrument='AIA', detector='AIA', measurement='171'` to work out the sourceid inside of `download_jp2`.

Reply:
Yeah we work it out inside download_jp2 but we do need to pass it to API. For that we need it to be a parameter or if we use kwargs to pass them then we have to make it compulsory that kwargs must contain source ID. Moreover if the user types in this command
'filepath = hv.download_jp2('2012/07/03 14:30:00', observatory='SDO', instrument='AIA', detector='AIA', measurement='171')'
then he will call the function which will call the API with wrong parameters.

Reply:
We download it using two different ways
1. 'filepath = hv.download_jp2('2012/07/03 14:30:00', observatory='SDO', instrument='AIA', detector='AIA', measurement='171')'

2. file = hv.download_jp2('2012/07/03 14:30:00', sourceId=data_sources['SOHO']['LASCO']['C2']['white-light']['sourceId'])

We can use the 2nd method but we cannot use the first method

Reply:
Yes, we need to change the sunpy API of `download_jp2` but I don't understand what that has to do with passing sourceid into the helioviewer API.